### DISCUSSION TITLE
micropython.native gone in 1.21 ?

### DISCUSSION THREAD

**Discussion by zcattacz** [micropython/micropython#12854] [ESP32] [answered]
Updated a ESP32-S2 to 1.21 firmware, and my previous CRC calc code ceased working. there was a small `@native` annotated portion.

```
 >>> sys.implementation
(name='micropython', version=(1, 21, 0), _machine='LOLIN_S2_MINI with ESP32-S2FN4R2', _mpy=10502)
 >>> from micropython import native
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: can't import name native
```

Just wondering, is this function removed from 1.21.0 ?

**Comment by jimmo** [comment]
>  Just wondering, is this function removed from 1.21.0 ?

The code generation decorators (`native`, `viper`, `asm_xtensa`, etc) do not actually exist in the `micropython` module. Rather the compiler detects them explicitly and handles them. See https://github.com/micropython/micropython/issues/11805#issuecomment-1596392266

The native decorator definitely still works in v1.21. I just tested it on an S2.

Can you share the code you're using?

**Reply by zcattacz** [reply]
Thank you, I got it.

I must have tried to make the code with `@native` inside compatible in cpython without actually testing it in 1.20.
```
try:
    from micropython import native  
except
    def native(fn):
        return fn
```

Is there any trick to let cpython ignore @micropython.native?

**Reply by jimmo** [reply, marked as answer]
> Is there any trick to let cpython ignore @micropython.native?

The common thing to do here is to write a micropython.py that you use with CPython (but don't copy it to your device) that contains dummy decorators.

**Reply by zcattacz** [reply]
Hum, silly me. thanks.
